### ISSUE TITLE
Crashes Word and WordPad when trying to set the font

### ISSUE THREAD

**Issue by DownLoude** [dscorbett/duployan-font#4]
I tried to use the font to write 𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆 in Word and WordPad and they both crashed. WordPad works for small glyphs but crashes for longer sentences, Word crashes every time. WordPad also does not render the the glyphs properly.

**Comment by dscorbett** [comment]
This font was designed to work with [HarfBuzz](https://harfbuzz.github.io/). I’m not surprised that it doesn’t work in Microsoft’s shaping engine. With that said, it is a long-term goal for the font to support all shaping engines, even if some are not as well-supported as HarfBuzz. For example, it might fall back to showing each letter separately, as in “𛰜‌𛰃‌𛱇‌𛰚‌𛱄‌𛰊‌𛰋‌𛱁‌𛰄‌𛱆”, or it might look mostly right but with adjacent words sometimes overlapping each other.

Please post a screenshot of the bad rendering in WordPad.

What kind of Duployan text do you want to write? Isolated short words and phrases, or full paragraphs? Any particular language? Would showing each letter separately be good enough or do you need proper cursive connections?

**Comment by DownLoude** [comment]
**I've looked at HarfBuzz, but I'm not really sure what I'm doing with it, the documentation is so technical**. It doesn't seem like a standalone application so I assume it is implemented somewhere? I saw something with LibreOffice, but that doesn't display it right either (I'm probably missing something).
(LibreOffice)
![image](https://user-images.githubusercontent.com/49001257/191358629-8473a269-8f95-410f-8bf2-641a12773a86.png)

I have an interest for obscure writing systems, and Duployan is one I've stumbled upon while researching the script just before it in Unicode, Nüshu. **I want to use it to write words for presentation, but also to transcribe documents (what I do for most scripts I document)**. I saw Noto Sans Duployan but it wasn't a joining font and doesn't reflect the beauty of Duployan. This is the first writing system that I've really looked into that isn't just character by character, but rather joined in a systematic way, so I'm not sure what I'm doing. 

**The font
![image](https://user-images.githubusercontent.com/49001257/191356002-484b3d49-345c-467b-a153-a8738330ae9a.png)
Same thing in Noto Sans Duployan
![image](https://user-images.githubusercontent.com/49001257/191356058-8d231d03-e606-4c07-a86e-1a771146e0d6.png)**

**Comment by dscorbett** [comment]
HarfBuzz is a library used within applications. Most browsers use it, which is why [the demo keyboard](https://dscorbett.github.io/duployan-test/) works. LibreOffice also uses it, but for some reason the font doesn’t work there. Your original string, “𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆”, looks like this in LibreOffice:
<img width="110" alt="𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆" src="https://user-images.githubusercontent.com/1124347/191369500-bdafcf3f-8a02-47b2-83b0-46ad029a6c75.png">
Every glyph is contextualized and oriented correctly, but they are positioned incorrectly.

According to your screenshots, WordPad doesn’t even do orientation correctly. It doesn’t seem to do anything. WordPad might be a lost cause. The best I can do is showing each glyph separately like Noto Sans Duployan does. That is low priority.

For now, the most practical way to use the font is in a browser, and the most practical way to share Duployan text is take screenshots of the rendering in your browser.

If you are planning to transcribe documents, be warned that the encoding implemented by this font is subject to change. Texts that look correct with the current version may look different in later versions. I am specifically thinking of vowel sequences. For example, “𛰃𛱁𛱆” <U+1BC03, U+1BC41, U+1BC46> looks like this now:
![𛰃𛱁𛱆](https://user-images.githubusercontent.com/1124347/191371328-feeb959b-6be4-4e7f-be7e-3cd49cd5e63f.png)
but may soon look like this:
![𛰃𛱁𛱆](https://user-images.githubusercontent.com/1124347/191373030-2f758821-18bc-42ae-8685-2beff411c5cc.png)

**Comment by DownLoude** [comment]
Yeah, I've been using the demo keyboard for most of my exploration so far, with the rest on paper. But just for ease of use, is there any text editor that can produce pdfs or something similar that works with this font that you know of? Is there something for TeX? I've only ever used TeXworks so idk if that works.

If not then thanks, you've been super helpful.

**Comment by khaledhosny** [comment]
Where do I get the font, I see no downloads. I’d like to at least check what is going on with LibreOffice.

**Comment by dscorbett** [comment]
https://github.com/dscorbett/duployan-test/blob/gh-pages/assets/fonts/Duployan-Regular.otf

**Comment by khaledhosny** [comment]
I get the same output from LibreOffice and hb-view:
<img width="845" alt="image" src="https://user-images.githubusercontent.com/93914/192086336-a343d56b-c490-4d1a-a0ce-bdb4d8dbdb7f.png">

**Comment by dscorbett** [comment]
That string is the one I posted to simulate a fallback rendering with separated letters. How does this issue’s original string, “𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆”, look? I just upgraded to LibreOffice 7.4.1.2 and it is now rendered thus for me:
<img width="110" alt="𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆" src="https://user-images.githubusercontent.com/1124347/192101482-bad01a41-a485-4189-ac89-461c4642ddab.png">
That is nearly right, but the cursive anchor points seem to be slightly offset. The problem disappears at font sizes, or if I zoom in. My former version of LibreOffice must have been using a too old HarfBuzz.

**Comment by dscorbett** [comment]
Anyway, @khaledhosny, thank you for testing.

@DownLoude, you could try exporting PDFs with a recent version of LibreOffice.

**Comment by khaledhosny** [comment]
> That is nearly right, but the cursive anchor points seem to be slightly offset. The problem disappears at font sizes, or if I zoom in.

Yes, this is a general issue due to aggressive rounding of glyph positions. It might get better in later 7.4 releases or may be in 7.5 (there have been some recent work in this area but I haven’t been paying close attention).

**Comment by khaledhosny** [comment]
> How does this issue’s original string, “𛰜𛰃𛱇𛰚‌𛰃𛱄͏͏͏‌𛰊𛰋𛱂‌𛰄𛱆”, look?

<img width="125" alt="image" src="https://user-images.githubusercontent.com/93914/192102722-641e5767-71ac-4adf-957c-e7df5aa8c2b1.png">
